### PR TITLE
Week2

### PR DESCRIPTION
**Pull request by salar01**
eplaced callbacks by promises, and edited week1 according to Jim's comments.

### CODE REVIEW

File: homework/src/index.js [hunk 1]
```diff
@@ -1,27 +1,29 @@
...
   }
 
   function createAndAppend(name, parent, options = {}) {
     const elem = document.createElement(name);
     parent.appendChild(elem);
-    Object.keys(options).forEach((key) => {
+    Object.keys(options).forEach(key => {
       const value = options[key];
-      if (key === 'html') {
+      if (key === "text") {
```
Comment (JawharBairakdar) [inline homework/src/index.js]: I really Liked this change :smiley:

File: homework/src/index.js [hunk 2]
```diff
@@ -31,17 +33,116 @@
...
+    const tbody = createAndAppend("tbody", table);
+    const tr = createAndAppend("tr", tbody);
+    const container = createAndAppend("div", root, { class: "container" });
+    createAndAppend("td", tr, { text: "Repository: ", class: "label" });
+    const repositoryLink = createAndAppend("a", tbody);
+    const description = createAndAppend("tr", tbody);
+    const forks = createAndAppend("tr", tbody, { class: "label" });
+
+    const updated = createAndAppend("tr", tbody, { class: "label" });
+
```
Comment (JawharBairakdar) [inline homework/src/index.js]: By looking to the [documentations](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/tbody) under the example table you will find a blue table that contains the Permitted content header aside, there you can find the valid values you can use.

Until this poit you have HTML generated:
```javascript
<table class="table">
  <tbody>
    <tr>
      <td class="label">Repository: </td>
    </tr>
    <a></a>
    <tr></tr>
    <tr class="label"></tr>
    <tr class="label"></tr>
  </tbody>
</table>
```
and by validating with [w3c validator](https://validator.w3.org/nu/#textarea) your code we will notice that you have here:
screenshot_2018-09-04 showing results for contents of text-input area - nu html checker

So in order to fix this the only thing you have to do is just to move your `a` tag into a `td` that is inside of a `tr` tag so you will have the following:
```javascript
<table class="table">
  <tbody>
    <tr>
      <td class="label">Repository: </td>
    </tr>
    <tr>
      <td>
        <a></a> <!-- Like so -->      
      </td>
    </tr>
    <tr></tr>
    <tr class="label"></tr>
    <tr class="label"></tr>
  </tbody>
</table>
```

File: homework/src/index.js [hunk 3]
```diff
@@ -31,17 +33,116 @@
...
+    fetchJSON(url)
+      .catch(error => {
```
Comment (JawharBairakdar) [inline homework/src/index.js]: catch or then first ? I think it's very important question and here is the [answer](https://youtu.be/M2R2pkKv1k8?t=1054) I think you watched it before but re-watching things we did watched them before is important after applying it sometimes :slightly_smiling_face:

File: homework/src/index.js [hunk 4]
```diff
@@ -31,17 +33,116 @@
...
+        createAndAppend("div", root, {
+          text: error.message,
+          class: "alert-error"
+        });
+        return;
```
Comment (JawharBairakdar) [inline homework/src/index.js]: No need for the `return` keyword here, the `return` keyword is used usually to return something or to stop the function from executing any thing else, while you have it in the last line of your callback function it just meaningless for the function for stopping it and for you if you need to return something from there.

File: homework/src/index.js [hunk 5]
```diff
@@ -31,17 +33,116 @@
...
+
+        renderRepositoryBox(
+          repositories[0],
+          repositoryLink,
+          forks,
+          description,
+          updated
+        );
+
+        select.addEventListener("change", () => {
```
Comment (JawharBairakdar) [inline homework/src/index.js]: this line is telling me that you did understand the callback concept :slightly_smiling_face:.

File: homework/src/index.js [hunk 6]
```diff
@@ -31,17 +33,116 @@
...
+          forks,
+          description,
+          updated
+        ) {
+          repositoryLink.innerText = repositories.name;
+          repositoryLink.setAttribute("href", repositories.html_url);
+          repositoryLink.setAttribute("target", "_blank");
+
+          forks.innerHTML = "Forks: " + repositories.forks;
+          description.innerHTML = "Description: " + repositories.description;
```
Comment (JawharBairakdar) [inline homework/src/index.js]: you sre using too much `innerHTML` and `setAttribute` while you can create your tags here and for each of them you can set the `text` as your change in `createAndAppend` function right?

File: homework/src/index.js [hunk 7]
```diff
@@ -31,17 +33,116 @@
...
+          updated.innerHTML =
+            "Updated: " + new Date(repositories.updated_at).toLocaleString();
+        };
+
+        renderContributors(container, repositories[0].contributorsUrl,);
```
Comment (JawharBairakdar) [inline homework/src/index.js]: If you followed the last comment you will have the container is launched from this function, and then you can pass it via argument.

File: homework/src/index.js [hunk 8]
```diff
@@ -31,17 +33,116 @@
...
+          src, element:avatar_url,
+          
+        });
+        
+        
+        createAndAppend("h3", li, {
+          text: element.contributions,
+          class: "contributors_badge"
+        });
+      });
```
Comment (JawharBairakdar) [inline homework/src/index.js]: Till this point You have:
```javascript
<div class="container">
  <ul class="right_div">
    Contributions: 
    <li></li>
  </ul>
</div>
```
it's not really done yet but you may make it better by thinking about it this way:
I will fetch the repositories and for each repository I will fetch contributers and then I will render each contributers.

**Comment by salar01** [comment]
Hello Jawhar, i am really impressed by your comments and spend lot of times to look through my homework. I really apprciate you. I wanna one more time go through whole JavaScript course agian, actually after consulting with Gijs, i will do that. So i stopped with class16 and i follow this course with class17 from 23 September(JavaScrip2) again. I hope i will use your guidence with class17 and also through my study with whole concept of JavaScript again. Thanks for everything. :) 👍 

PS : I will study JavaScript again, and i finish this and previous homeworks also :)

**Comment by JawharBairakdar** [comment]
Sounds Great :slightly_smiling_face:, see you in class 17.
Best of luck

**Pull request closed without merging**